src/common/io.cc

```diff
@@ -6,14 +6,96 @@
 #include <fcntl.h>
 #include <unistd.h>
 #endif  // defined(__unix__)
+#include <algorithm>
 #include <cstdio>
 #include <string>
 
 #include "xgboost/logging.h"
+#include "io.h"
 
 namespace xgboost {
 namespace common {
 
+size_t PeekableInStream::Read(void* dptr, size_t size) {
+  size_t nbuffer = buffer_.length() - buffer_ptr_;
+  if (nbuffer == 0) return strm_->Read(dptr, size);
+  if (nbuffer < size) {
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_) + buffer_ptr_, nbuffer);
+    buffer_ptr_ += nbuffer;
+    return nbuffer + strm_->Read(reinterpret_cast<char*>(dptr) + nbuffer,
+                                 size - nbuffer);
+  } else {
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_) + buffer_ptr_, size);
+    buffer_ptr_ += size;
+    return size;
+  }
+}
+
+size_t PeekableInStream::PeekRead(void* dptr, size_t size) {
+  size_t nbuffer = buffer_.length() - buffer_ptr_;
+  if (nbuffer < size) {
+    buffer_ = buffer_.substr(buffer_ptr_, buffer_.length());
+    buffer_ptr_ = 0;
+    buffer_.resize(size);
+    size_t nadd = strm_->Read(dmlc::BeginPtr(buffer_) + nbuffer, size - nbuffer);
+    buffer_.resize(nbuffer + nadd);
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_), buffer_.length());
+    return buffer_.length();
+  } else {
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_) + buffer_ptr_, size);
+    return size;
+  }
+}
+
+FixedSizeStream::FixedSizeStream(PeekableInStream* stream) : PeekableInStream(stream), pointer_{0} {
+  size_t size {1}, total {0};
```

Comment:
Should we set the initial size to some bigger number like 32 or 16? I believe a lot of allocators do not allocate less then 8 bytes anyway so we won't actually save any memory by doing .resize(1) or .resize(2) but we would incur 4-5 additional calls

Reply:
@trams Let me think about it.  This fixed stream might not be necessary.